cryostatio/cryostat-agent
# feat(harvester): implement JFR harvester and push-to-archives

Depends on https://github.com/cryostatio/cryostat-core/pull/169
Depends on #18

Related to https://github.com/cryostatio/cryostat/issues/520
Related to https://github.com/cryostatio/cryostat/issues/1013

Fixes #22
Related to #23 - there is still experimentation and work to be done to expose controls for the emergency dump data age and filesize
Fixes #24
See also #25

How to test: usual `-agent` test setup in brief
1. Check out PR, `mvn clean install` to get agent JAR into local `~/.m2`
2. Check out latest `quarkus-test`, `./mvnw -U clean package && podman build -t quay.io/andrewazores/quarkus-test:latest -f src/main/docker/Dockerfile.jvm . ; podman image prune -f` to build a `quarkus-test` with the PR agent included
3. `cd cryostat`, edit one of the `quarkus-test-agent` sample apps to use `:latest` tag from image above
4. `sh smoketest.sh`

After that, the standard `smoketest.sh` should be running with Cryostat working as usual, three `vertx-fib-demo` as usual, and two `quarkus-test-agent` - one with the PR version agent and one with the `main` version.

`podman logs -f quarkus-test-agent-1` (if `-1` is the one with the PR agent) to check what the new agent says it's doing. With no additional configuration done, the new and old agents should behave as usual - only as discovery plugins, with no harvester behaviour. They should be visible in the Cryostat API in all the usual ways.

Edit `smoketest.sh`:


```diff
diff --git a/smoketest.sh b/smoketest.sh
index 48c0f88c..28937c59 100755
--- a/smoketest.sh
+++ b/smoketest.sh
@@ -126,6 +126,7 @@ runDemoApps() {
         --env QUARKUS_HTTP_PORT=10010 \
         --env ORG_ACME_CRYOSTATSERVICE_ENABLED="false" \
         --env CRYOSTAT_AGENT_APP_NAME="quarkus-test-agent" \
+        --env CRYOSTAT_AGENT_HARVESTER_PERIOD_MS=30000 \
         --env CRYOSTAT_AGENT_WEBSERVER_HOST="localhost" \
         --env CRYOSTAT_AGENT_WEBSERVER_PORT="9977" \
         --env CRYOSTAT_AGENT_CALLBACK="http://localhost:9977/" \
```

to enable the harvester behaviour on one of the `quarkus-test` sample apps, and re-run `smoketest.sh`. This configures the agent to start a recording (defaults to using the `default.jfc` shipped with the host JVM) and to push the collected data to the Cryostat backend every 30 seconds. These should be visible in the Cryostat UI as they should produce upload notifications, and should be listed in the `Archives > Uploads` view.
